Sources/SwiftInspectorVisitors/TypeDescription.swift

```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  private var caseDescription: String {
+    switch self {
+    case .array:
+      return Self.arrayDescription
+    case .composition:
+      return Self.compositionDescription
+    case .dictionary:
+      return Self.dictionaryDescription
+    case .implicitlyUnwrapped:
+      return Self.implicitlyUnwrappedDescription
+    case .member:
+      return Self.memberDescription
+    case .optional:
+      return Self.optionalDescription
+    case .simple:
+      return Self.simpleDescription
+    case .tuple:
+      return Self.tupleDescription
+    case .unknown:
+      return Self.unknownDescription
+    }
+  }
+
+  private static let simpleDescription = "simple"
+  private static let memberDescription = "member"
+  private static let compositionDescription = "composition"
+  private static let optionalDescription = "optional"
+  private static let implicitlyUnwrappedDescription = "implicitlyUnwrapped"
+  private static let arrayDescription = "array"
+  private static let dictionaryDescription = "dictionary"
+  private static let tupleDescription = "tuple"
+  private static let unknownDescription = "unknown"
+}
+
+extension TypeSyntax {
+
+  /// Returns the type description for the receiver.
+  /// - Warning: Do not call on a type syntax node of type `ClassRestrictionTypeSyntax`,
+  ///            `SomeTypeSyntax`, `MetatypeTypeSyntax`, `FunctionTypeSyntax`,
+  ///            `AttributedTypeSyntax`, or `UnknownTypeSyntax`
+  var typeDescription: TypeDescription {
+    if let typeIdentifier = self.as(SimpleTypeIdentifierSyntax.self) {
+      return .simple(name: typeIdentifier.name.text)
+
+    } else if let typeIdentifier = self.as(MemberTypeIdentifierSyntax.self) {
+      return .member(
+        name: typeIdentifier.name.text,
+        baseType: typeIdentifier.baseType.typeDescription)
+
+    } else if let typeIdentifiers = self.as(CompositionTypeSyntax.self) {
+      return .composition(typeIdentifiers.elements.map { $0.type.typeDescription })
+
+    } else if let typeIdentifier = self.as(OptionalTypeSyntax.self) {
+      return .optional(typeIdentifier.wrappedType.typeDescription)
+
+    } else if let typeIdentifier = self.as(ImplicitlyUnwrappedOptionalTypeSyntax.self) {
+      return .implicitlyUnwrapped(typeIdentifier.wrappedType.typeDescription)
+
+    } else if let typeIdentifier = self.as(ArrayTypeSyntax.self) {
+      return .array(typeIdentifier.elementType.typeDescription)
+
+    } else if let typeIdentifier = self.as(DictionaryTypeSyntax.self) {
+      return .dictionary(
+        key: typeIdentifier.keyType.typeDescription,
+        value: typeIdentifier.valueType.typeDescription)
+
+    } else if let typeIdentifiers = self.as(TupleTypeSyntax.self) {
+      return .tuple(typeIdentifiers.elements.map { $0.type.typeDescription })
+
+    } else if self.as(ClassRestrictionTypeSyntax.self) != nil {
+      // A class restriction is the same as requiring inheriting from AnyObject.
+      return .simple(name: "AnyObject")
```

Comment:
Note that I'm being tricky here. `protocol SomeObject: class` is functionally equivalent to `protocol SomeObject: AnyObject`, so I'm just pretending that this is the latter rather than creating a new enum case to handle this somewhat deprecated way to restrict a protocol to a reference type.

Reply:
For the code review gods, can we link to some source showing that they are equivalent?

Reply:
https://forums.swift.org/t/class-only-protocols-class-vs-anyobject/11507/4

Reply:
b84bcfb953dfce5e0cd0592b042d9976d914fdaa